packages/base-driver/lib/basedriver/commands/find.js

```diff
@@ -1,51 +1,108 @@
-const commands = {}, helpers = {}, extensions = {};
+/* eslint-disable no-unused-vars */
+/* eslint-disable require-await */
+// @ts-check
+import {errors} from '../../protocol';
 
-
-// Override the following function for your own driver, and the rest is taken
-// care of!
-
-// helpers.findElOrEls = async function (strategy, selector, mult, context) {}
-//   strategy: locator strategy
-//   selector: the actual selector for finding an element
-//   mult: multiple elements or just one?
-//   context: finding an element from the root context? or starting from another element
-//
-// Returns an object which adheres to the way the JSON Wire Protocol represents elements:
-// { ELEMENT: # }    eg: { ELEMENT: 3 }  or { ELEMENT: 1.023 }
-
-helpers.findElOrElsWithProcessing = async function findElOrElsWithProcessing (strategy, selector, mult, context) {
-  this.validateLocatorStrategy(strategy);
-  try {
-    return await this.findElOrEls(strategy, selector, mult, context);
-  } catch (err) {
-    if (this.opts.printPageSourceOnFindFailure) {
-      const src = await this.getPageSource();
-      this.log.debug(`Error finding element${mult ? 's' : ''}: ${err.message}`);
-      this.log.debug(`Page source requested through 'printPageSourceOnFindFailure':`);
-      this.log.debug(src);
+/**
+ *
+ * @param {ReturnType<import('./event').EventMixin>} Base
```

Comment:
oh wow so we have to apply these in order?

Reply:
Sort of.  It'd be one thing if these were _pure_ mixins, but they aren't.  They don't work standalone; they call out to other methods in `DriverCore` or maybe even other mixin classes.  You can't just apply them to any given class.  

It sucks, but I couldn't find another way to do this without making things so generic we don't know what's in the final class.  TS doesn't handle our use case very well.

If none of these actually call out to other mixins (I can't recall), then the order itself isn't important, but the types still are.  It's rather arbitrary right now.

Reply:
another avenue I explored was just trying to trick TS into thinking that `BaseDriver` had a bunch of methods that were implemented elsewhere.  declaration merging (e.g., defining an `interface BaseDriver`) didn't work because TS generated broken declarations. module augmentation (`declare module './driver' { }`) didn't work because you cannot "add" stuff to a typed class.  so idk.  I was pulling my hair out.

all of this goes away, of course, if we just make a huge class.

Reply:
also, nothing in `DriverCore` or `BaseDriverCore` is allowed to call out to a method defined in any of these mixins.  that--and the discrepancies introduced in `AppiumDriver` was why I had to move stuff around.

Reply:
if typescript understood prototypes at all (it doesn't), I think this would be way easier.  but it's fundamentally based on classical OO and inheritance and all that garbage.  it makes some things difficult, and we found one of them.

```ts
class Foo {}

Foo.prototype = {bar: () => 'baz'};

let foo = new Foo();
foo.bar(); // TS238974239 error wtf no such method `bar`
```